app/src/main/java/com/google/android/sambadocumentsprovider/SambaConfiguration.java

```diff
@@ -45,42 +45,74 @@
   private static final String CONF_KEY_VALUE_SEPARATOR = " = ";
 
   private final File mHomeFolder;
+  private final File mShareFolder;
   private final Map<String, String> mConfigurations = new HashMap<>();
 
-  SambaConfiguration(File homeFolder) {
+  SambaConfiguration(File homeFolder, @Nullable File shareFolder) {
     mHomeFolder = homeFolder;
+    if (shareFolder != null && (shareFolder.isDirectory() || shareFolder.mkdirs())) {
+      mShareFolder = shareFolder;
+    } else {
+      Log.w(TAG, "Failed to create share folder. Only default value is supported.");
+
+      // Use home folder as the share folder to avoid null checks everywhere.
+      mShareFolder = mHomeFolder;
+    }
 
     setHomeEnv(homeFolder.getAbsolutePath());
   }
 
-  public void flushAsDefault() {
+  void flushDefault(OnConfigurationChangedListener listener) {
+    // lmhosts are not used in SambaDocumentsProvider and prioritize bcast because sometimes in home
+    // settings DNS will resolve unknown domain name to a specific IP for advertisement.
+    //
+    // lmhosts -- lmhosts file if existed side by side to smb.conf
+    // wins -- Windows Internet Name Service
+    // hosts -- hosts file and DNS resolution
+    // bcast -- NetBIOS broadcast
+    addConfiguration("name resolve order", "wins bcast hosts");
+
+    // Urge from users to disable SMB1 by default.
+    addConfiguration("client min protocol", "SMB2");
+    addConfiguration("client max protocol", "SMB3");
+
     File smbFile = getSmbFile(mHomeFolder);
     if (!smbFile.exists()) {
-      flush(null);
+      flush(listener);
     }
   }
 
-  public synchronized SambaConfiguration addConfiguration(String key, String value) {
+  synchronized SambaConfiguration addConfiguration(String key, String value) {
     mConfigurations.put(key, value);
     return this;
   }
 
-  public synchronized SambaConfiguration removeConfiguration(String key) {
+  synchronized SambaConfiguration removeConfiguration(String key) {
     mConfigurations.remove(key);
     return this;
   }
 
-  public void load(OnConfigurationChangedListener listener) {
-    new LoadTask(listener).execute();
+  boolean syncFromExternal(OnConfigurationChangedListener listener) {
+    final File smbFile = getSmbFile(mHomeFolder);
+    final File extSmbFile = getExtSmbFile(mShareFolder);
+
+    if (extSmbFile.isFile() && extSmbFile.lastModified() > smbFile.lastModified()) {
```

Comment:
is there any possibility that smbFile doesn't exist yet?  or will last modified just be 0 if no file?